### PR TITLE
Bluetooth: GATT: Various optimizations

### PR DESCRIPTION
**Pull request by Vudentz**
This adds optimization for handle lookup into the database and persist the database hash in flash so it can be used to check if service changed needs to be emitted on reboot.

### CODE REVIEW

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -933,6 +933,15 @@ void bt_gatt_foreach_attr(u16_t start_handle, u16_t end_handle,
 
 	SYS_SLIST_FOR_EACH_CONTAINER(&db, svc, node) {
 		int i;
+		struct bt_gatt_service *next;
+
+		next = SYS_SLIST_PEEK_NEXT_CONTAINER(svc, node);
+		if (next) {
+			/* Skip ahead if start is not within service handles */
+			if (next->attrs[0].handle <= start_handle) {
```
Comment (jhedberg) [inline subsys/bluetooth/host/gatt.c]: Should this really be <= and not < ?

Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: It is the next service so equal still means is not in the current svc.

Comment (jhedberg) [reply subsys/bluetooth/host/gatt.c]: right

**Review by jhedberg — APPROVED** [review]

File: subsys/bluetooth/host/gatt.c (hunk 1 above)
Comment (carlescufi) [reply subsys/bluetooth/host/gatt.c]: `next->attrs[0].handle` is the handle of the service definition of the next service? If so I don't get why you would skip ahead in that case

Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: Service, and its attributes, are registered in ascending order, so this is checking if the handle happens to be >= the following service which means is not withing the current service so it just skip iterating into the service handles. It is allowed to set a fixed handle, provided it is not in use, which is why we dont simple use handle + num_handles logic, though the logic is quite similar.

File: subsys/bluetooth/host/settings.c [hunk 2]
```diff
@@ -89,20 +89,15 @@ int bt_settings_decode_key(char *key, bt_addr_le_t *addr)
...
-		const struct bt_settings_handler *h;
-
-		for (h = _bt_settings_start; h < _bt_settings_end; h++) {
-			if (!strcmp(argv[0], h->name)) {
-				argc--;
-				argv++;
+	for (h = _bt_settings_start; h < _bt_settings_end; h++) {
+		if (!strcmp(argv[0], h->name)) {
+			argc--;
+			argv++;
```
Comment (jhedberg) [inline subsys/bluetooth/host/settings.c]: I guess this change to the generic bt settings code should be split into its own commit with an explanation of what/why.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/settings.c (hunk 2 above)
Comment (Vudentz) [reply subsys/bluetooth/host/settings.c]: Alright, it seems the top level (bt/) was restricted to just keys settings so this is changing to allow settings also only on top level.

Comment (Vudentz) [reply subsys/bluetooth/host/settings.c]: Actually my last comment is not quite right, the problem seems to be that we always expect an extra argument when in fact there could be settings that don't require one.

Comment (Vudentz) [reply subsys/bluetooth/host/settings.c]: It is now a separate patch.

**Review by jhedberg — APPROVED** [review]
